# Introducción a Machine Learning - Lorenzo Tomás Diez

## Actividad 1

### Situación

El RMS Titanic fue, en su momento, el mayor barco de pasajeros del mundo. Se hundió en su viaje inaugural de Southampton a Nueva York, en el año 1912. En el evento, fallecieron 1514 de las 2223 personas que iban a bordo, entre tripulación y pasajeros.

Hoy, en el año 2022, se ha decidido hacer un estudio de machine learning en el cual se requiere de nuestras labores como técnicos en Data Science, para identificar diversos patrones que permitan verificar si, basándonos en el entrenamiento de nuestro modelo de datos, la máquina es capaz de predecir si una persona logra sobrevivir o no.

Para ello, se utilizará una base muestral de 1220 datos, que forman parte del conjunto de datos, que figura en el archivo adjunto "Titanic_M1.xlsx".

### Consignas

Para realizar los futuros modelos de datos, el primer desafío será preprocesar la base de datos, según las siguientes instrucciones:

a) Describir brevemente la cantidad de valores faltantes para cada una de las variables de la base de datos.

b) En la columna Pclass, completar los valores NA, asignando a las personas a la clase 2.

c) En la clase "Sex", imputar los datos de manera de poder completar la base solo con los valores "male" y "female".

d) En los casos de las personas de las cuales no se tenga la edad, rellenar los datos con el promedio de edad de aquellos que sí tienen el dato.

e) En la columna embarked, asignar el valor "S" a aquellos datos que se encuentran con valor vacío.

### Formato de entrega:

#### ¿Cómo se debe presentar el trabajo?

La entrega debe ser mediante un archivo "ipynb" (formato de extensión de notebook para Python).

Cada una de las preguntas debe contener los códigos de cómo se lograron resolver las consignas solicitadas.

### Entrega

¡Llegaste al final de la actividad de este módulo! Recuerda guardar tus respuestas y luego subirlas clicando en el botón "Enviar tarea".

Puedes consultar tus dudas con tus compañeros en el foro de la materia o con tu tutor.

In [35]:
import pandas as pd

In [36]:
df = pd.read_excel('titanic_M1.xlsx')
df.head()

pclass  survived                                             name     sex  \
0     1.0         1                    Allen, Miss. Elisabeth Walton  female   
1     1.0         1                   Allison, Master. Hudson Trevor    male   
2     1.0         0                     Allison, Miss. Helen Loraine  female   
3     1.0         0             Allison, Mr. Hudson Joshua Creighton    male   
4     1.0         0  Allison, Mrs. Hudson J C (Bessie Waldo Daniels)  female   

       age  sibsp  parch  ticket       fare    cabin embarked boat   body  \
0  29.0000      0      0   24160  2113375.0       B5        S    2    NaN   
1   0.9167      1      2  113781  1515500.0  C22 C26        S   11    NaN   
2   2.0000      1      2  113781  1515500.0  C22 C26        S  NaN    NaN   
3  30.0000      1      2  113781  1515500.0  C22 C26        S  NaN  135.0   
4  25.0000      1      2  113781  1515500.0  C22 C26        S  NaN    NaN   

                         home.dest  
0                     St Louis, MO  
1  Montreal, PQ / Chesterville, ON  
2  Montreal, PQ / Chesterville, ON  
3  Montreal, PQ / Chesterville, ON  
4  Montreal, PQ / Chesterville, ON

In [37]:
df.dtypes

pclass       float64
survived       int64
name          object
sex           object
age          float64
sibsp          int64
parch          int64
ticket        object
fare         float64
cabin         object
embarked      object
boat          object
body         float64
home.dest     object
dtype: object

In [38]:
missing_values = df.isnull().sum()
missing_values

pclass         76
survived        0
name            0
sex             0
age           239
sibsp           0
parch           0
ticket          0
fare            0
cabin         927
embarked        2
boat          751
body         1110
home.dest     475
dtype: int64

In [39]:
df['pclass'] = df['pclass'].fillna(2)
df.isnull().sum()

pclass          0
survived        0
name            0
sex             0
age           239
sibsp           0
parch           0
ticket          0
fare            0
cabin         927
embarked        2
boat          751
body         1110
home.dest     475
dtype: int64

In [40]:
df['sex'].unique()

array(['female', 'male', 'Hombre', 'Mujer'], dtype=object)

In [41]:
df['sex'] = df['sex'].replace('Hombre', 'male')
df['sex'] = df['sex'].replace('Mujer', 'female')
df['sex'].unique()

array(['female', 'male'], dtype=object)

In [42]:
df['age'].unique()

array([29.    ,  0.9167,  2.    , 30.    , 25.    , 48.    , 63.    ,
       39.    , 53.    , 71.    , 47.    , 18.    , 24.    , 26.    ,
       80.    ,     nan, 50.    , 32.    , 36.    , 37.    , 42.    ,
       19.    , 35.    , 28.    , 45.    , 40.    , 58.    , 22.    ,
       41.    , 44.    , 59.    , 60.    , 33.    , 17.    , 11.    ,
       14.    , 49.    , 76.    , 46.    , 27.    , 64.    , 55.    ,
       70.    , 38.    , 51.    , 31.    ,  4.    , 54.    , 23.    ,
       43.    , 52.    , 16.    , 32.5   , 21.    , 15.    , 65.    ,
       28.5   , 45.5   , 56.    , 13.    , 61.    , 34.    ,  6.    ,
       57.    , 62.    , 67.    ,  1.    , 12.    , 20.    ,  0.8333,
        8.    ,  0.6667,  7.    ,  3.    , 36.5   , 18.5   ,  5.    ,
       66.    ,  9.    ,  0.75  , 70.5   , 22.5   ,  0.3333,  0.1667,
       40.5   , 10.    , 23.5   , 34.5   , 20.5   , 30.5   , 55.5   ,
       38.5   , 14.5   , 24.5   ])

In [43]:
df_ages_not_null = df[df['age'].notnull()]
average_age = df_ages_not_null['age'].mean()
average_age

30.02627551020408

In [44]:
df['age'] = df['age'].fillna(average_age)
df.isnull().sum()

pclass          0
survived        0
name            0
sex             0
age             0
sibsp           0
parch           0
ticket          0
fare            0
cabin         927
embarked        2
boat          751
body         1110
home.dest     475
dtype: int64

In [45]:
df['embarked'] = df['embarked'].fillna('S')
df.isnull().sum()

pclass          0
survived        0
name            0
sex             0
age             0
sibsp           0
parch           0
ticket          0
fare            0
cabin         927
embarked        0
boat          751
body         1110
home.dest     475
dtype: int64

## Resolucion
### A - Describir brevemente la cantidad de valores faltantes para cada una de las variables de la base de datos.
En la base de datos, la cantidad de valores faltantes por variable es la siguiente:

- pclass: 76 valores faltantes.
- age: 239 valores faltantes.
- cabin: 927 valores faltantes.
- embarked: 2 valores faltantes.
- boat: 751 valores faltantes.
- body: 1110 valores faltantes.
- home.dest: 475 valores faltantes.

### B - En la columna Pclass, completar los valores NA, asignando a las personas a la clase 2.

```
df['pclass'] = df['pclass'].fillna(2)
df.isnull().sum()
```

### C - En la clase "Sex", imputar los datos de manera de poder completar la base solo con los valores "male" y "female".
```
df['sex'] = df['sex'].replace('Hombre', 'male')
df['sex'] = df['sex'].replace('Mujer', 'female')
df['sex'].unique()
```

### D - En los casos de las personas de las cuales no se tenga la edad, rellenar los datos con el promedio de edad de aquellos que sí tienen el dato.
```
df_ages_not_null = df[df['age'].notnull()]
average_age = df_ages_not_null['age'].mean()
df['age'] = df['age'].fillna(average_age)
```

### E - En la columna embarked, asignar el valor "S" a aquellos datos que se encuentran con valor vacío.
```
df['embarked'] = df['embarked'].fillna('S')
```